### PR TITLE
Refactor go planner

### PR DESCRIPTION
**Pull request by nevermore3**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

#### Issue(s) number: 
close https://github.com/vesoft-inc/nebula/issues/5295

#### Description:
`yield 'Tim Duncan' as a | go 3 steps from $-.a over like where like.likeness > 30 yield edge as e`

previous  plan

image

current plan
image

Due to the existence of the loop operator in the previous plan, the data flow and dependencies were inconsistent, so many optimization rules could not be applied, such as filter pushdown, merging projects, and so on.

in the new plan.  the go statement is divided into two operators, expand and expandall, expand is responsible for expansion, and expandAll is responsible for the expansion process to obtain the attributes required by the user

BEFORE:

```
(root@nebula) [nba]> yield 'Tim Duncan' AS id| go 100 steps from $-.id  over * bidirect yield edge as e|yield count(*)
+----------+
| count(*) |
+----------+
| 486      |
+----------+
Got 1 rows (time spent 3178857/3188150 us)
```

NOW:
```
(root@nebula) [nba]> yield 'Tim Duncan' AS id| go 100 steps from $-.id  over * bidirect yield edge as e|yield count(*)
+----------+
| count(*) |
+----------+
| 486      |
+----------+
Got 1 rows (time spent 209215/209876 us)
```

image

the yellow line is the optimized version
the blue line is the master version

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: tests/tck/features/go/GoYieldVertexEdge.feature [hunk 1]
```diff
@@ -54,13 +54,6 @@ Feature: Go Yield Vertex And Edge Sentence
       | ("LaMarcus Aldridge" :player{age: 33, name: "LaMarcus Aldridge"}) | "teammate" |
       | ("Manu Ginobili" :player{age: 41, name: "Manu Ginobili"})         | "teammate" |
       | ("Tony Parker" :player{age: 36, name: "Tony Parker"})             | "teammate" |
-      | ("Manu Ginobili" :player{age: 41, name: "Manu Ginobili"})         | "like"     |
```
Comment (nevermore3) [inline tests/tck/features/go/GoYieldVertexEdge.feature:57]: After discussing with @MuYiYong  offline, the incompatibility of this situation is acceptable

File: src/graph/executor/query/ExpandAllExecutor.cpp [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+namespace graph {
+
+Status ExpandAllExecutor::buildRequestVids() {
+  SCOPED_TIMER(&execTime_);
+  const auto& inputVar = expand_->inputVar();
+  auto inputIter = ectx_->getResult(inputVar).iterRef();
+  auto iter = static_cast<SequentialIter*>(inputIter);
+  size_t iterSize = iter->size();
+  nextStepVids_.reserve(iterSize);
+  QueryExpressionContext ctx(ectx_);
```
Comment (yixinglu) [inline src/graph/executor/query/ExpandAllExecutor.cpp:26]: useless

File: src/graph/executor/query/ExpandAllExecutor.cpp [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+        memory::MemoryCheckGuard guard;
+        nextStepVids_.clear();
+        SCOPED_TIMER(&execTime_);
+        addStats(resp);
+        time::Duration expandTime;
+        curLimit_ = 0;
+        curMaxLimit_ = stepLimits_.empty() ? std::numeric_limits<int64_t>::max()
+                                           : stepLimits_[currentStep_ - 2];
+        return handleResponse(std::move(resp)).ensure([this, expandTime]() {
+          std::string timeName = "graphExpandExpandTime+" + folly::to<std::string>(currentStep_);
```
Comment (yixinglu) [inline src/graph/executor/query/ExpandAllExecutor.cpp:105]: "graphExpandAllTime"?

File: src/graph/executor/query/ExpandAllExecutor.cpp [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+          otherStats_.emplace(timeName, folly::sformat("{}(us)", expandTime.elapsedInUSec()));
```
Comment (yixinglu) [inline src/graph/executor/query/ExpandAllExecutor.cpp:106]: addState(timeName, expandTime.elapsedInUSec());

File: src/graph/executor/query/ExpandAllExecutor.cpp [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+          if (!nextStepVids_.empty()) {
+            return getNeighbors();
+          } else if (!preVisitedVids_.empty()) {
+            return expandFromCache();
+          } else {
+            return finish(ResultBuilder().value(Value(std::move(result_))).build());
+          }
+        } else {
+          return finish(ResultBuilder().value(Value(std::move(result_))).build());
+        }
```
Comment (yixinglu) [inline src/graph/executor/query/ExpandAllExecutor.cpp:124]:
```c++
        if (currentStep_ <= maxSteps_) {
          if (!nextStepVids_.empty()) {
            return getNeighbors();
          } 
          if (!preVisitedVids_.empty()) {
            return expandFromCache();
          }
        }
        return finish(ResultBuilder().value(Value(std::move(result_))).build());
```

File: src/graph/executor/query/ExpandAllExecutor.cpp [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+      std::sort(samples.begin(), samples.end(), [](int64_t a, int64_t b) { return a > b; });
+    }
+
+    std::unordered_map<Value, std::unordered_set<Value>> dst2VidsMap;
+    std::unordered_set<Value> visitedVids;
+    getNeighborsFromCache(dst2VidsMap, visitedVids, samples);
+    preVisitedVids_.swap(visitedVids);
+    preDst2VidsMap_.swap(dst2VidsMap);
+    std::string timeName = "graphCacheExpandTime+" + folly::to<std::string>(currentStep_);
+    otherStats_.emplace(timeName, folly::sformat("{}(us)", expandTime.elapsedInUSec()));
```
Comment (yixinglu) [inline src/graph/executor/query/ExpandAllExecutor.cpp:155]: see above comment

File: src/graph/executor/query/ExpandAllExecutor.cpp [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+    std::unordered_set<Value>& visitedVids,
+    std::vector<int64_t>& samples) {
+  for (const auto& vid : preVisitedVids_) {
+    auto findVid = adjList_.find(vid);
+    if (findVid == adjList_.end()) {
+      continue;
+    }
+    auto& adjEdgeProps = findVid->second;
+    auto& vertexProps = adjEdgeProps.back();
+    for (auto edgeIter = adjEdgeProps.begin(); edgeIter != adjEdgeProps.end() - 1; ++edgeIter) {
```
Comment (yixinglu) [inline src/graph/executor/query/ExpandAllExecutor.cpp:174]: edgeIter < adjEdgeProps.end()

File: src/graph/executor/query/ExpandAllExecutor.cpp [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+  }
+  auto vidIter = nextStepVids_.begin();
+  while (vidIter != nextStepVids_.end()) {
+    if (adjList_.find(*vidIter) != adjList_.end()) {
+      visitedVids.emplace(*vidIter);
+      vidIter = nextStepVids_.erase(vidIter);
+    } else {
+      vidIter++;
+    }
+  }
```
Comment (yixinglu) [inline src/graph/executor/query/ExpandAllExecutor.cpp:321]: could you extract the same logics into the functions for the usage in `handleResponse` and `getNeighborsFromCache`?

File: tests/tck/features/optimizer/PushSampleDownRule.feature [hunk 9]
```diff
@@ -1,12 +1,13 @@
 # Copyright (c) 2021 vesoft inc. All rights reserved.
 #
 # This source code is licensed under Apache 2.0 License.
+@jmq
```
Comment (yixinglu) [inline tests/tck/features/optimizer/PushSampleDownRule.feature:4]: cleanup

File: src/graph/executor/query/ExpandAllExecutor.cpp (hunk 8 above)
Comment (nevermore3) [reply src/graph/executor/query/ExpandAllExecutor.cpp]: done

File: src/graph/executor/query/ExpandAllExecutor.cpp (hunk 7 above)
Comment (nevermore3) [reply src/graph/executor/query/ExpandAllExecutor.cpp:171]: should be edgeIter < adjEdgeProps.end() - 1 ,  and  edgeIter != adjEdgeProps.end() -1 is ok

**Review by yixinglu — DISMISSED** [review]
LGTM

**Review by jievince — COMMENTED** [review]
Big change! Looking forward a benchmark

**Review by yixinglu — APPROVED** [review]

**Review by jievince — APPROVED** [review]

**Pull request merged**